File: python3/validate_stack_sequences.py
"""
Space   : O(n)
Time    : O(n)
"""


class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        n = len(pushed)
        stack = []
        i, j = 0, 0

        while i < n or stack:
            while stack and stack[-1] == popped[j]:
                stack.pop()
                j += 1
            
            if i < n:
                stack.append(pushed[i])
                i += 1
            else:
                break

        return len(stack) == 0

File: golang/validate_stack_sequences.go
package main

// Space   : O(n)
// Time    : O(n)

func validateStackSequences(pushed []int, popped []int) bool {
	n := len(pushed)
	var stack []int
	i, j := 0, 0

	for i < n || len(stack) > 0 {
		if len(stack) > 0 && stack[len(stack)-1] == popped[j] {
			stack = stack[:len(stack)-1]
			j++
		} else if i < n {
			stack = append(stack, pushed[i])
			i++
		} else {
			return false
		}
	}
	return true
}
